### ISSUE TITLE
Why is partitionCount required for proper partitioning?

### ISSUE THREAD

**Issue by elnur** [spring-cloud/spring-cloud-stream#977]
Can't the system figure out the number of partitions on the fly? Doesn't the high level Kafka producer partition properly without specifying the number of partitions explicitly?

I believe that unless the number of partitions is specified explicitly, it should default to introspecting the topic to get its number of partitions on the fly.

**Comment by mbogoevici** [comment]
@elnur Here are some answers:

>Can't the system figure out the number of partitions on the fly? Doesn't the high level Kafka producer partition properly without specifying the number of partitions explicitly?

Depends on the messaging system. You're asking a question about Kafka, but this is a property that potentially applies to other messaging systems, that do not support partitions natively (e.g. RabbitMQ http://docs.spring.io/autorepo/docs/spring-cloud-stream-docs/Chelsea.SR2/reference/htmlsingle/#_rabbitmq_binder_overview. ). 

In the case of Kafka, the next answer applies :). 

> I believe that unless the number of partitions is specified explicitly, it should default to introspecting the topic to get its number of partitions on the fly.

This is what the Kafka binder does. Notice this in particular: https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/blob/master/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java#L161

Thanks for looking into Spring Cloud Stream and for reaching out!

**Comment by elnur** [comment]
@mbogoevici, thanks for getting back.

Yeah, it looks like that code is supposed to solve the problem, but that's not what I observed when testing it.

I had a topic with 12 partitions and I didn't configure `partitionCount`. I configured `partitionKeyExpression` though. The observed behavior was that messages with different partition keys went into the same first partition. As soon as I set `partitionCount` to 12, it all started working properly — that is, all messages with the same partition key ended up in the same partition and message with different partition keys ended up in different partitions.

Am I missing something? Is there a configuration property I need to set to make this work properly without setting `partitionCount` explicitly?
